File: src/utils/costUtils.ts
import type { Subscription, SpendSummary } from '../types';

function toDaysInCycle(sub: Subscription): number {
  const amount = sub.customCycleAmount ?? 1;
  const unit = sub.customCycleUnit ?? 'months';
  if (unit === 'days') return amount;
  if (unit === 'weeks') return amount * 7;
  return amount * 30.4375; // average days per month
}

export function toMonthly(sub: Subscription): number {
  switch (sub.billingCycle) {
    case 'weekly':
      return sub.cost * (52 / 12);
    case 'monthly':
      return sub.cost;
    case 'yearly':
      return sub.cost / 12;
    case 'custom': {
      const daysInCycle = toDaysInCycle(sub);
      return sub.cost * (365.25 / daysInCycle / 12);
    }
  }
}

export function toYearly(sub: Subscription): number {
  return toMonthly(sub) * 12;
}

export function computeSpendSummary(subs: Subscription[]): SpendSummary {
  const currencies = new Set(subs.map((s) => s.currency));
  return {
    monthly: subs.reduce((acc, s) => acc + toMonthly(s), 0),
    yearly: subs.reduce((acc, s) => acc + toYearly(s), 0),
    currency: currencies.size === 1 ? [...currencies][0] : 'Mixed',
  };
}

export function formatCurrency(amount: number, currency: string): string {
  if (currency === 'Mixed') {
    return `~$${amount.toFixed(2)}`;
  }
  try {
    return new Intl.NumberFormat(undefined, {
      style: 'currency',
      currency,
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }).format(amount);
  } catch {
    return `${currency} ${amount.toFixed(2)}`;
  }
}

export function formatSubscriptionCost(sub: Subscription): string {
  const cycleLabel: Record<string, string> = {
    weekly: '/week',
    monthly: '/mo',
    yearly: '/yr',
    custom: '',
  };
  const suffix = sub.billingCycle === 'custom'
    ? `/${sub.customCycleAmount ?? 1} ${sub.customCycleUnit ?? 'months'}`
    : cycleLabel[sub.billingCycle];
  return `${formatCurrency(sub.cost, sub.currency)}${suffix}`;
}

File: android-app/app/src/main/kotlin/com/subhandler/utils/CostUtils.kt
package com.subhandler.utils

import com.subhandler.model.BillingCycle
import com.subhandler.model.CustomCycleUnit
import com.subhandler.model.SpendSummary
import com.subhandler.model.Subscription
import java.text.NumberFormat
import java.util.Currency
import java.util.Locale
import kotlin.math.roundToInt

fun toMonthly(subscription: Subscription): Double = when (subscription.billingCycle) {
    BillingCycle.WEEKLY  -> subscription.cost * (52.0 / 12.0)
    BillingCycle.MONTHLY -> subscription.cost
    BillingCycle.YEARLY  -> subscription.cost / 12.0
    BillingCycle.CUSTOM  -> {
        val amount = subscription.customCycleAmount ?: 1
        val daysInCycle = when (subscription.customCycleUnit ?: CustomCycleUnit.MONTHS) {
            CustomCycleUnit.DAYS   -> amount.toDouble()
            CustomCycleUnit.WEEKS  -> amount * 7.0
            CustomCycleUnit.MONTHS -> amount * 30.4375
        }
        subscription.cost * (365.25 / daysInCycle / 12.0)
    }
}

fun toYearly(subscription: Subscription): Double = toMonthly(subscription) * 12.0

fun computeSpendSummary(subscriptions: List<Subscription>): SpendSummary {
    if (subscriptions.isEmpty()) return SpendSummary()
    val monthly = subscriptions.sumOf { toMonthly(it) }
    val yearly = subscriptions.sumOf { toYearly(it) }
    val currencies = subscriptions.map { it.currency }.toSet()
    val currency = if (currencies.size == 1) currencies.first() else "Mixed"
    return SpendSummary(monthly = monthly, yearly = yearly, currency = currency, count = subscriptions.size)
}

fun formatCurrency(amount: Double, currency: String): String {
    if (currency == "Mixed") {
        return "~${"%.2f".format(amount)}"
    }
    return try {
        val format = NumberFormat.getCurrencyInstance(Locale.US)
        format.currency = Currency.getInstance(currency)
        format.maximumFractionDigits = 2
        format.minimumFractionDigits = 2
        format.format(amount)
    } catch (e: Exception) {
        "$currency ${"%.2f".format(amount)}"
    }
}

fun formatSubscriptionCost(subscription: Subscription): String {
    val costStr = "%.2f".format(subscription.cost)
    val cycleLabel = formatCycleLabel(subscription)
    return "${subscription.currency} $costStr$cycleLabel"
}
